core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
> What I'm troubled with, is how to state metadata requirements.

A project like this interfaces with the real world and with all manner of software which we cannot control, not to mention conditions which we cannot control - like "the network is down", "the database is down", "Tokyo has fallen into the sea", and an unknown subset of metadata is unavailable at that time. But when someone types "ls" there still has to be a response.

The point is that if you don't plan for failure, you have already failed. Until we have a clear view of what is to be gained by imposing more rigid requirements, I think it ought to be something like: as a producer, provide whatever metadata you wish. As a consumer, degrade gracefully if the metadata you were hoping for isn't there.